File: GottaGo-Android/app/src/main/java/com/gottago/app/service/ToiletDataService.kt
package com.gottago.app.service

import com.google.gson.Gson
import com.google.gson.JsonArray
import com.google.gson.JsonObject
import com.gottago.app.model.*
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.net.URLEncoder
import java.util.concurrent.TimeUnit

object ToiletDataService {

    private val client = OkHttpClient.Builder()
        .connectTimeout(20, TimeUnit.SECONDS)
        .readTimeout(35, TimeUnit.SECONDS)
        .build()
    private val gson = Gson()

    // ──────────────────────────────────────────────────────────────────────────
    // Overpass / OpenStreetMap
    // ──────────────────────────────────────────────────────────────────────────

    suspend fun fetchOverpassToilets(
        lat: Double, lng: Double, radiusDegrees: Double = 0.045
    ): List<Bathroom> = withContext(Dispatchers.IO) {
        val south = lat - radiusDegrees; val north = lat + radiusDegrees
        val west  = lng - radiusDegrees; val east  = lng + radiusDegrees
        val bb    = "$south,$west,$north,$east"

        val query = """
            [out:json][timeout:30];
            (
              node["amenity"="toilets"]($bb);
              way["amenity"="toilets"]($bb);
              node["toilets"="yes"]($bb);
              way["toilets"="yes"]($bb);
            );
            out center 100;
        """.trimIndent()

        val body = ("data=" + URLEncoder.encode(query, "UTF-8"))
            .toRequestBody("application/x-www-form-urlencoded".toMediaType())
        val request = Request.Builder()
            .url("https://overpass-api.de/api/interpreter")
            .post(body)
            .build()

        try {
            val response = client.newCall(request).execute()
            val json = gson.fromJson(response.body?.string(), JsonObject::class.java)
            val elements = json?.getAsJsonArray("elements") ?: return@withContext emptyList()
            elements.mapNotNull { overpassToBathroom(it.asJsonObject) }
        } catch (e: Exception) {
            println("Overpass error: ${e.message}")
            emptyList()
        }
    }

    private fun overpassToBathroom(el: JsonObject): Bathroom? {
        val id = el["id"]?.asLong ?: return null

        val lat: Double
        val lng: Double
        if (el.has("lat") && el.has("lon")) {
            lat = el["lat"].asDouble; lng = el["lon"].asDouble
        } else {
            val center = el["center"]?.asJsonObject ?: return null
            lat = center["lat"].asDouble; lng = center["lon"].asDouble
        }

        val tags = el["tags"]?.asJsonObject
        val amenity = tags?.get("amenity")?.asString ?: ""
        val isDedicated = amenity == "toilets"

        val (type, requiresPurchase) = if (isDedicated) {
            BathroomType.PUBLIC_FACILITY to false
        } else {
            osmVenueType(tags)
        }

        val name = if (isDedicated) {
            tags?.get("name")?.asString ?: tags?.get("description")?.asString ?: "Public Toilet"
        } else {
            val venueName = tags?.get("name")?.asString
            if (!venueName.isNullOrBlank()) "$venueName – Restroom"
            else "${type.displayName} Restroom"
        }

        val fee = when (tags?.get("fee")?.asString) {
            "no", "free" -> FeeType.FREE
            "yes"        -> FeeType.PAID
            else         -> if (isDedicated) FeeType.UNKNOWN else FeeType.FREE
        }
        val accessible    = tags?.get("wheelchair")?.asString.let { it == "yes" || it == "designated" }
        val genderNeutral = tags?.get("unisex")?.asString == "yes"
        val hours         = tags?.get("opening_hours")?.asString ?: ""
        val access        = tags?.get("access")?.asString ?: ""
        var notes         = if (access.isEmpty()) "" else "Access: $access"
        if (requiresPurchase) notes = if (notes.isEmpty()) "May require a purchase." else "$notes May require a purchase."

        val idHex   = id.toString(16).padStart(12, '0').uppercase()
        val uuidStr = "4F534D30-0000-4000-8000-$idHex"

        return Bathroom(
            id = uuidStr,
            name = name,
            address = "",
            latitude = lat,
            longitude = lng,
            type = type,
            fee = fee,
            isAccessible = accessible,
            isGenderNeutral = genderNeutral,
            requiresPurchase = requiresPurchase,
            notes = notes,
            hours = hours,
            isVerified = true
        )
    }

    private fun osmVenueType(tags: JsonObject?): Pair<BathroomType, Boolean> {
        val amenity = tags?.get("amenity")?.asString ?: ""
        val tourism  = tags?.get("tourism")?.asString ?: ""
        val shop     = tags?.get("shop")?.asString ?: ""
        val leisure  = tags?.get("leisure")?.asString ?: ""

        return when (amenity) {
            "fuel"                                                              -> BathroomType.GAS_STATION to false
            "restaurant","fast_food","cafe","bar","pub","food_court","biergarten","ice_cream"
                                                                                -> BathroomType.RESTAURANT  to true
            "hospital","clinic","doctors","pharmacy","dentist",
            "library","community_centre","arts_centre","theatre","cinema",
            "ferry_terminal","bus_station","airport"                            -> BathroomType.PUBLIC_FACILITY to false
            "shopping_mall"                                                     -> BathroomType.STORE to false
            "supermarket","convenience"                                         -> BathroomType.STORE to true
            else -> when (tourism) {
                "hotel","motel","hostel","guest_house","apartment","chalet","camp_site"
                    -> BathroomType.HOTEL to false
                else -> when {
                    shop.isNotEmpty() -> BathroomType.STORE to true
                    leisure in listOf("park","recreation_ground","garden","nature_reserve","dog_park","playground")
                        -> BathroomType.PARK to false
                    else -> BathroomType.OTHER to false
                }
            }
        }
    }

    // ──────────────────────────────────────────────────────────────────────────
    // Refuge Restrooms
    // ──────────────────────────────────────────────────────────────────────────

    suspend fun fetchRefugeRestrooms(lat: Double, lng: Double): List<Bathroom> =
        withContext(Dispatchers.IO) {
            val url = "https://www.refugerestrooms.org/api/v1/restrooms/by_location.json" +
                "?lat=$lat&lng=$lng&per_page=20&accessible=false"
            val request = Request.Builder().url(url).build()
            try {
                val response = client.newCall(request).execute()
                val arr = gson.fromJson(response.body?.string(), JsonArray::class.java) ?: return@withContext emptyList()
                arr.mapNotNull { refugeToBathroom(it.asJsonObject) }
            } catch (e: Exception) {
                println("Refuge error: ${e.message}")
                emptyList()
            }
        }

    private fun refugeToBathroom(obj: JsonObject): Bathroom? {
        val bLat = obj["latitude"]?.asDouble?.takeIf { it != 0.0 } ?: return null
        val bLng = obj["longitude"]?.asDouble?.takeIf { it != 0.0 } ?: return null
        val id   = obj["id"]?.asInt ?: 0
        val idHex   = Math.abs(id).toLong().toString(16).padStart(12, '0').uppercase()
        val uuidStr = "52465547-0000-4000-8000-$idHex"

        val rawName = obj["name"]?.asString?.trim() ?: ""
        val name    = rawName.ifEmpty { "Refuge Restroom" }
        val address = listOfNotNull(
            obj["street"]?.asString?.trim()?.ifEmpty { null },
            obj["city"]?.asString?.trim()?.ifEmpty { null },
            obj["state"]?.asString?.trim()?.ifEmpty { null },
            obj["country"]?.asString?.trim()?.ifEmpty { null }
        ).joinToString(", ")

        return Bathroom(
            id = uuidStr,
            name = name,
            address = address,
            latitude = bLat,
            longitude = bLng,
            fee = FeeType.FREE,
            isAccessible = obj["accessible"]?.asBoolean ?: false,
            isGenderNeutral = true,
            notes = obj["directions"]?.asString ?: "",
            isVerified = true
        )
    }

    // ──────────────────────────────────────────────────────────────────────────
    // Great British Toilet Map (GraphQL)
    // ──────────────────────────────────────────────────────────────────────────

    suspend fun fetchGBPTM(lat: Double, lng: Double, radiusMeters: Int = 5000): List<Bathroom> =
        withContext(Dispatchers.IO) {
            val bodyJson = """{"query":"query NearbyLoos(${'$'}lat:Float!,${'$'}lng:Float!,${'$'}radius:Int!){loosByProximity(lat:${'$'}lat,lng:${'$'}lng,radius:${'$'}radius){id name accessible allGender noPayment openingTimes removalReason location{lat lng}}}","variables":{"lat":$lat,"lng":$lng,"radius":$radiusMeters}}"""
            val body = bodyJson.toRequestBody("application/json".toMediaType())
            val request = Request.Builder()
                .url("https://api.toiletmap.org.uk/graphql")
                .post(body)
                .build()
            try {
                val response = client.newCall(request).execute()
                val json = gson.fromJson(response.body?.string(), JsonObject::class.java)
                val loos = json?.getAsJsonObject("data")
                    ?.getAsJsonArray("loosByProximity") ?: return@withContext emptyList()
                loos.mapNotNull { gbptmToBathroom(it.asJsonObject) }
            } catch (e: Exception) {
                println("GBPTM error: ${e.message}")
                emptyList()
            }
        }

    private fun gbptmToBathroom(obj: JsonObject): Bathroom? {
        if (!obj["removalReason"]?.isJsonNull!!) return null // skip removed loos
        val loc = obj["location"]?.asJsonObject ?: return null
        val bLat = loc["lat"]?.asDouble ?: return null
        val bLng = loc["lng"]?.asDouble ?: return null
        val rawId   = obj["id"]?.asString ?: ""
        val hexPart = rawId.filter { it.isLetterOrDigit() }.take(12).padEnd(12, '0').uppercase()
        val uuidStr = "47425054-0000-4000-8000-$hexPart"
        val name    = obj["name"]?.asString?.trim()?.ifEmpty { null } ?: "Public Toilet"
        val noPayment = obj["noPayment"]?.takeIf { !it.isJsonNull }?.asBoolean
        val fee = when (noPayment) {
            true  -> FeeType.FREE
            false -> FeeType.PAID
            null  -> FeeType.UNKNOWN
        }
        return Bathroom(
            id = uuidStr,
            name = name,
            latitude = bLat,
            longitude = bLng,
            fee = fee,
            isAccessible = obj["accessible"]?.takeIf { !it.isJsonNull }?.asBoolean ?: false,
            isGenderNeutral = obj["allGender"]?.takeIf { !it.isJsonNull }?.asBoolean ?: false,
            hours = obj["openingTimes"]?.takeIf { !it.isJsonNull }?.asString ?: "",
            isVerified = true
        )
    }

    // ──────────────────────────────────────────────────────────────────────────
    // Australia National Toilet Map
    // ──────────────────────────────────────────────────────────────────────────

    suspend fun fetchAustraliaToilets(lat: Double, lng: Double, maxResults: Int = 50): List<Bathroom> =
        withContext(Dispatchers.IO) {
            val url = "https://www.toiletmap.gov.au/Home/GetNearestToilets" +
                "?Latitude=$lat&Longitude=$lng&MaxNumber=$maxResults&Radius=5"
            val request = Request.Builder()
                .url(url)
                .header("User-Agent", "Mozilla/5.0 (Linux; Android 14) AppleWebKit/537.36")
                .build()
            try {
                val response = client.newCall(request).execute()
                val arr = gson.fromJson(response.body?.string(), JsonArray::class.java) ?: return@withContext emptyList()
                arr.mapNotNull { ausToiletToBathroom(it.asJsonObject) }
            } catch (e: Exception) {
                println("Australia error: ${e.message}")
                emptyList()
            }
        }

    private fun ausToiletToBathroom(obj: JsonObject): Bathroom? {
        val bLat = obj["Latitude"]?.takeIf { !it.isJsonNull }?.asDouble?.takeIf { it != 0.0 } ?: return null
        val bLng = obj["Longitude"]?.takeIf { !it.isJsonNull }?.asDouble?.takeIf { it != 0.0 } ?: return null
        val toiletId = obj["ToiletID"]?.takeIf { !it.isJsonNull }?.asInt ?: 0
        val idHex   = Math.abs(toiletId).toLong().toString(16).padStart(12, '0').uppercase()
        val uuidStr = "41555354-0000-4000-8000-$idHex"

        val name = obj["Name"]?.takeIf { !it.isJsonNull }?.asString?.trim()?.ifEmpty { null } ?: "Public Toilet"
        val address = listOfNotNull(
            obj["Address1"]?.takeIf { !it.isJsonNull }?.asString?.trim()?.ifEmpty { null },
            obj["Town"]?.takeIf { !it.isJsonNull }?.asString?.trim()?.ifEmpty { null },
            obj["State"]?.takeIf { !it.isJsonNull }?.asString?.trim()?.ifEmpty { null },
            obj["Postcode"]?.takeIf { !it.isJsonNull }?.asString?.trim()?.ifEmpty { null }
        ).joinToString(", ")

        fun str(key: String) = obj[key]?.takeIf { !it.isJsonNull }?.asString ?: ""
        val accessible    = str("AccessibleMale") == "Yes" || str("AccessibleFemale") == "Yes" || str("AccessibleUnisex") == "Yes"
        val genderNeutral = str("Unisex") == "Yes"
        val fee           = if (str("PayToilet") == "Yes") FeeType.PAID else FeeType.FREE
        val notes         = buildList {
            if (str("BabyChange") == "Yes") add("Baby change available.")
            str("Notes").takeIf { it.isNotEmpty() }?.let { add(it) }
        }.joinToString(" ")

        return Bathroom(
            id = uuidStr,
            name = name,
            address = address,
            latitude = bLat,
            longitude = bLng,
            fee = fee,
            isAccessible = accessible,
            isGenderNeutral = genderNeutral,
            notes = notes,
            hours = str("OpeningHours"),
            isVerified = true
        )
    }

    // ──────────────────────────────────────────────────────────────────────────
    // Submit to Refuge Restrooms
    // ──────────────────────────────────────────────────────────────────────────

    suspend fun submitToRefugeRestrooms(bathroom: Bathroom): Boolean =
        withContext(Dispatchers.IO) {
            val parts   = bathroom.address.split(",").map { it.trim() }
            val payload = JsonObject().apply {
                addProperty("name",       bathroom.name)
                addProperty("street",     parts.getOrElse(0) { bathroom.address })
                addProperty("city",       parts.getOrElse(1) { "" })
                addProperty("state",      parts.getOrElse(2) { "" })
                addProperty("country",    parts.getOrElse(3) { "US" }.ifEmpty { "US" })
                addProperty("accessible", bathroom.isAccessible)
                addProperty("unisex",     bathroom.isGenderNeutral)
                addProperty("latitude",   bathroom.latitude)
                addProperty("longitude",  bathroom.longitude)
                addProperty("directions", bathroom.notes)
            }
            val body    = gson.toJson(payload).toRequestBody("application/json".toMediaType())
            val request = Request.Builder()
                .url("https://www.refugerestrooms.org/api/v1/restrooms")
                .post(body)
                .build()
            try {
                client.newCall(request).execute().code == 201
            } catch (e: Exception) {
                println("Refuge submit error: ${e.message}")
                false
            }
        }
}

File: GottaGo/ToiletDataService.swift
import Foundation
import CoreLocation

struct ToiletDataService {

    // MARK: - Overpass API (OpenStreetMap public toilets)

    static func fetchOverpassToilets(near coord: CLLocationCoordinate2D,
                                     radiusDegrees: Double = 0.045) async -> [Bathroom] {
        let south = coord.latitude  - radiusDegrees
        let north = coord.latitude  + radiusDegrees
        let west  = coord.longitude - radiusDegrees
        let east  = coord.longitude + radiusDegrees
        let bb = "\(south),\(west),\(north),\(east)"

        // Query 1: dedicated toilet facilities (highest priority)
        // Query 2: any venue (gas station, hotel, restaurant, shop, park…) tagged toilets=yes
        let query = """
        [out:json][timeout:30];
        (
          node["amenity"="toilets"](\(bb));
          way["amenity"="toilets"](\(bb));
          node["toilets"="yes"](\(bb));
          way["toilets"="yes"](\(bb));
        );
        out center 100;
        """

        guard let url = URL(string: "https://overpass-api.de/api/interpreter") else { return [] }
        var request = URLRequest(url: url)
        request.httpMethod  = "POST"
        request.httpBody    = ("data=" + (query.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? ""))
            .data(using: .utf8)
        request.timeoutInterval = 35

        do {
            let (data, _) = try await URLSession.shared.data(for: request)
            let response  = try JSONDecoder().decode(OverpassResponse.self, from: data)
            return response.elements.compactMap { overpassToBathroom($0) }
        } catch {
            print("Overpass fetch error: \(error.localizedDescription)")
            return []
        }
    }

    // MARK: - Refuge Restrooms API (gender-neutral / trans-friendly)

    static func fetchRefugeRestrooms(near coord: CLLocationCoordinate2D) async -> [Bathroom] {
        let urlStr = "https://www.refugerestrooms.org/api/v1/restrooms/by_location.json"
            + "?lat=\(coord.latitude)&lng=\(coord.longitude)&per_page=20&accessible=false"
        guard let url = URL(string: urlStr) else { return [] }

        do {
            let (data, _) = try await URLSession.shared.data(from: url)
            let entries   = try JSONDecoder().decode([RefugeEntry].self, from: data)
            return entries.compactMap { refugeToBathroom($0) }
        } catch {
            print("Refuge Restrooms fetch error: \(error.localizedDescription)")
            return []
        }
    }

    // MARK: - Great British Toilet Map (UK — 14,000+ facilities)

    static func fetchGBPTM(near coord: CLLocationCoordinate2D,
                            radiusMeters: Int = 5_000) async -> [Bathroom] {
        guard let url = URL(string: "https://api.toiletmap.org.uk/graphql") else { return [] }

        // GraphQL query — filter out removed loos (removalReason != null)
        let graphql: [String: Any] = [
            "query": """
            query NearbyLoos($lat: Float!, $lng: Float!, $radius: Int!) {
              loosByProximity(lat: $lat, lng: $lng, radius: $radius) {
                id name accessible allGender noPayment openingTimes removalReason
                location { lat lng }
              }
            }
            """,
            "variables": [
                "lat": coord.latitude,
                "lng": coord.longitude,
                "radius": radiusMeters
            ]
        ]

        guard let body = try? JSONSerialization.data(withJSONObject: graphql) else { return [] }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = body
        request.timeoutInterval = 20

        do {
            let (data, _) = try await URLSession.shared.data(for: request)
            let response  = try JSONDecoder().decode(GBPTMResponse.self, from: data)
            return (response.data?.loosByProximity ?? [])
                .filter { $0.removalReason == nil }       // only active loos
                .compactMap { gbptmToBathroom($0) }
        } catch {
            print("GBPTM fetch error: \(error.localizedDescription)")
            return []
        }
    }

    // MARK: - Australia National Public Toilet Map (17,000+ gov-verified)

    static func fetchAustraliaToilets(near coord: CLLocationCoordinate2D,
                                      maxResults: Int = 50) async -> [Bathroom] {
        // Official Australian Government toilet map API
        let urlStr = "https://www.toiletmap.gov.au/Home/GetNearestToilets"
            + "?Latitude=\(coord.latitude)"
            + "&Longitude=\(coord.longitude)"
            + "&MaxNumber=\(maxResults)"
            + "&Radius=5"   // km
        guard let url = URL(string: urlStr) else { return [] }

        var request = URLRequest(url: url)
        request.timeoutInterval = 20
        // The gov site requires a browser-like User-Agent
        request.setValue("Mozilla/5.0 (iPhone; CPU iPhone OS 17_0 like Mac OS X) AppleWebKit/605.1.15",
                         forHTTPHeaderField: "User-Agent")

        do {
            let (data, _) = try await URLSession.shared.data(for: request)
            let toilets   = try JSONDecoder().decode([AusToilet].self, from: data)
            return toilets.compactMap { ausToiletToBathroom($0) }
        } catch {
            print("Australia toilet fetch error: \(error.localizedDescription)")
            return []
        }
    }

    // MARK: - Refuge Restrooms: Submit new entry

    /// Submits a new restroom to the Refuge Restrooms community database.
    /// Returns `true` on HTTP 201 Created.
    static func submitToRefugeRestrooms(_ bathroom: Bathroom) async -> Bool {
        guard let url = URL(string: "https://www.refugerestrooms.org/api/v1/restrooms") else {
            return false
        }

        // Split "Street, City, State, Country" out of the combined address field.
        let parts = bathroom.address
            .components(separatedBy: ",")
            .map { $0.trimmingCharacters(in: .whitespaces) }
        let street  = parts.indices.contains(0) ? parts[0] : bathroom.address
        let city    = parts.indices.contains(1) ? parts[1] : ""
        let state   = parts.indices.contains(2) ? parts[2] : ""
        let country = parts.indices.contains(3) ? parts[3] : ""

        var commentParts: [String] = []
        if !bathroom.hours.isEmpty      { commentParts.append("Hours: \(bathroom.hours)") }
        if bathroom.requiresPurchase    { commentParts.append("Purchase required.") }
        if let code = bathroom.accessCode, !code.isEmpty {
            commentParts.append("Access code: \(code)")
        }

        let body: [String: Any] = [
            "name":           bathroom.name,
            "street":         street,
            "city":           city,
            "state":          state,
            "country":        country.isEmpty ? "US" : country,
            "accessible":     bathroom.isAccessible,
            "unisex":         bathroom.isGenderNeutral,
            "changing_table": false,
            "directions":     bathroom.notes,
            "comment":        commentParts.joined(separator: " "),
            "latitude":       bathroom.latitude,
            "longitude":      bathroom.longitude
        ]

        guard let httpBody = try? JSONSerialization.data(withJSONObject: body) else { return false }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = httpBody
        request.timeoutInterval = 20

        do {
            let (_, response) = try await URLSession.shared.data(for: request)
            return (response as? HTTPURLResponse)?.statusCode == 201
        } catch {
            print("Refuge Restrooms submit error: \(error.localizedDescription)")
            return false
        }
    }

    // MARK: - Converters

    private static func overpassToBathroom(_ el: OverpassElement) -> Bathroom? {
        let lat: Double
        let lon: Double

        if let elLat = el.lat, let elLon = el.lon {
            lat = elLat; lon = elLon
        } else if let center = el.center {
            lat = center.lat; lon = center.lon
        } else {
            return nil
        }

        let tags = el.tags ?? [:]
        let amenity = tags["amenity"] ?? ""
        let isDedicatedToilet = amenity == "toilets"

        // Map OSM tags → BathroomType + purchase requirement
        let (bathroomType, requiresPurchase): (BathroomType, Bool) = isDedicatedToilet
            ? (.publicFacility, false)
            : osmVenueType(tags: tags)

        // Build a descriptive name: venues get " – Restroom" suffix
        let name: String
        if isDedicatedToilet {
            name = tags["name"] ?? tags["description"] ?? "Public Toilet"
        } else {
            if let venueName = tags["name"], !venueName.trimmingCharacters(in: .whitespaces).isEmpty {
                name = "\(venueName) – Restroom"
            } else {
                name = "\(bathroomType.displayName) Restroom"
            }
        }

        let fee: FeeType = {
            switch tags["fee"] {
            case "no", "free": return .free
            case "yes":        return .paid
            default:           return isDedicatedToilet ? .unknown : .free
            }
        }()

        let accessible    = tags["wheelchair"] == "yes" || tags["wheelchair"] == "designated"
        let genderNeutral = tags["unisex"] == "yes"
        let hours         = tags["opening_hours"] ?? ""
        let access        = tags["access"] ?? ""
        var notes         = access.isEmpty ? "" : "Access: \(access)"
        if requiresPurchase {
            notes = notes.isEmpty ? "May require a purchase." : notes + " May require a purchase."
        }

        // Deterministic UUID: prefix "4F534D30" = "OSM0" in ASCII hex
        let idHex   = String(format: "%012X", el.id)
        let uuidStr = "4F534D30-0000-4000-8000-\(idHex)"
        let id      = UUID(uuidString: uuidStr) ?? UUID()

        return Bathroom(
            id: id,
            name: name,
            address: "",
            latitude: lat,
            longitude: lon,
            type: bathroomType,
            fee: fee,
            isAccessible: accessible,
            isGenderNeutral: genderNeutral,
            requiresPurchase: requiresPurchase,
            accessCode: nil,
            notes: notes,
            hours: hours,
            isVerified: true,
            dateAdded: Date(),
            isFavorite: false,
            reviews: []
        )
    }

    /// Maps OSM tags to the best-fit BathroomType and whether a purchase is likely required.
    private static func osmVenueType(tags: [String: String]) -> (BathroomType, Bool) {
        let amenity = tags["amenity"] ?? ""
        let tourism  = tags["tourism"] ?? ""
        let shop     = tags["shop"] ?? ""
        let leisure  = tags["leisure"] ?? ""

        switch amenity {
        // Gas / fuel stations
        case "fuel":
            return (.gasStation, false)
        // Food & drink — toilet access often requires purchase
        case "restaurant", "fast_food", "cafe", "bar",
             "pub", "food_court", "biergarten", "ice_cream":
            return (.restaurant, true)
        // Public-service buildings — free access
        case "hospital", "clinic", "doctors", "pharmacy", "dentist",
             "library", "community_centre", "arts_centre",
             "theatre", "cinema", "place_of_worship",
             "ferry_terminal", "bus_station", "airport":
            return (.publicFacility, false)
        // Shopping complexes
        case "shopping_mall":
            return (.store, false)
        case "supermarket", "convenience":
            return (.store, true)
        default:
            break
        }
        // Accommodation
        switch tourism {
        case "hotel", "motel", "hostel", "guest_house",
             "apartment", "chalet", "camp_site":
            return (.hotel, false)
        default:
            break
        }
        // Retail shops
        if !shop.isEmpty { return (.store, true) }
        // Outdoor / leisure
        switch leisure {
        case "park", "recreation_ground", "garden", "nature_reserve",
             "dog_park", "playground":
            return (.park, false)
        case "sports_centre", "stadium", "fitness_centre",
             "swimming_pool", "golf_course":
            return (.publicFacility, false)
        default:
            break
        }
        return (.other, false)
    }

    private static func refugeToBathroom(_ entry: RefugeEntry) -> Bathroom? {
        guard entry.latitude != 0, entry.longitude != 0 else { return nil }

        // Deterministic UUID: prefix "52465547" = "RFUG" in ASCII hex
        let idHex   = String(format: "%012X", abs(entry.id))
        let uuidStr = "52465547-0000-4000-8000-\(idHex)"
        let id = UUID(uuidString: uuidStr) ?? UUID()

        let name    = entry.name.trimmingCharacters(in: .whitespaces).isEmpty
                      ? "Refuge Restroom"
                      : entry.name
        let address = [entry.street, entry.city, entry.state, entry.country]
            .compactMap { s -> String? in
                guard let s, !s.trimmingCharacters(in: .whitespaces).isEmpty else { return nil }
                return s
            }
            .joined(separator: ", ")

        return Bathroom(
            id: id,
            name: name,
            address: address,
            latitude: entry.latitude,
            longitude: entry.longitude,
            type: .publicFacility,
            fee: .free,
            isAccessible: entry.accessible,
            isGenderNeutral: true,
            requiresPurchase: false,
            accessCode: nil,
            notes: entry.directions ?? "",
            hours: "",
            isVerified: true,
            dateAdded: Date(),
            isFavorite: false,
            reviews: []
        )
    }

    private static func gbptmToBathroom(_ loo: GBPTMLoo) -> Bathroom? {
        guard let loc = loo.location else { return nil }

        // Deterministic UUID: prefix "47425054" = "GBPT" in ASCII hex
        // GBPTM IDs are hex strings — take first 12 hex chars
        let hexChars   = loo.id.filter { $0.isHexDigit }
        let hexPart    = String(hexChars.prefix(12)).padding(toLength: 12, withPad: "0", startingAt: 0).uppercased()
        let id         = UUID(uuidString: "47425054-0000-4000-8000-\(hexPart)") ?? UUID()
        let name       = loo.name?.trimmingCharacters(in: .whitespaces).isEmpty == false
                         ? loo.name! : "Public Toilet"
        let fee: FeeType = loo.noPayment == true ? .free : loo.noPayment == false ? .paid : .unknown

        return Bathroom(
            id: id,
            name: name,
            address: "",
            latitude: loc.lat,
            longitude: loc.lng,
            type: .publicFacility,
            fee: fee,
            isAccessible: loo.accessible ?? false,
            isGenderNeutral: loo.allGender ?? false,
            requiresPurchase: false,
            accessCode: nil,
            notes: "",
            hours: loo.openingTimes ?? "",
            isVerified: true,
            dateAdded: Date(),
            isFavorite: false,
            reviews: []
        )
    }

    private static func ausToiletToBathroom(_ t: AusToilet) -> Bathroom? {
        guard let lat = t.Latitude, let lon = t.Longitude,
              lat != 0, lon != 0 else { return nil }

        // Deterministic UUID: prefix "41555354" = "AUST" in ASCII hex
        let idHex   = String(format: "%012X", abs(t.ToiletID ?? 0))
        let id      = UUID(uuidString: "41555354-0000-4000-8000-\(idHex)") ?? UUID()
        let name    = t.Name?.isEmpty == false ? t.Name! : "Public Toilet"
        let parts   = [t.Address1, t.Town, t.State, t.Postcode]
            .compactMap { s -> String? in
                guard let s = s, !s.trimmingCharacters(in: .whitespaces).isEmpty else { return nil }
                return s
            }
        let address = parts.joined(separator: ", ")

        let accessible = (t.AccessibleMale == "Yes" ||
                          t.AccessibleFemale == "Yes" ||
                          t.AccessibleUnisex == "Yes")
        let genderNeutral = t.Unisex == "Yes"
        let fee: FeeType  = t.PayToilet == "Yes" ? .paid : .free

        var notes: [String] = []
        if t.BabyChange == "Yes" { notes.append("Baby change available.") }
        if let n = t.Notes, !n.isEmpty { notes.append(n) }

        return Bathroom(
            id: id,
            name: name,
            address: address,
            latitude: lat,
            longitude: lon,
            type: .publicFacility,
            fee: fee,
            isAccessible: accessible,
            isGenderNeutral: genderNeutral,
            requiresPurchase: false,
            accessCode: nil,
            notes: notes.joined(separator: " "),
            hours: t.OpeningHours ?? "",
            isVerified: true,
            dateAdded: Date(),
            isFavorite: false,
            reviews: []
        )
    }
}

// MARK: - JSON models

private struct OverpassResponse: Decodable {
    let elements: [OverpassElement]
}

private struct OverpassElement: Decodable {
    let id:     Int
    let lat:    Double?
    let lon:    Double?
    let center: OverpassCenter?
    let tags:   [String: String]?
}

private struct OverpassCenter: Decodable {
    let lat: Double
    let lon: Double
}

// MARK: - Great British Toilet Map JSON models

private struct GBPTMResponse: Decodable {
    let data: GBPTMData?
}
private struct GBPTMData: Decodable {
    let loosByProximity: [GBPTMLoo]?
}
private struct GBPTMLoo: Decodable {
    let id:            String
    let name:          String?
    let location:      GBPTMLocation?
    let accessible:    Bool?
    let allGender:     Bool?
    let noPayment:     Bool?
    let openingTimes:  String?
    let removalReason: String?
}
private struct GBPTMLocation: Decodable {
    let lat: Double
    let lng: Double
}

// MARK: - Australia National Toilet Map JSON model

private struct AusToilet: Decodable {
    let ToiletID:         Int?
    let Name:             String?
    let Address1:         String?
    let Town:             String?
    let State:            String?
    let Postcode:         String?
    let Latitude:         Double?
    let Longitude:        Double?
    let AccessibleMale:   String?
    let AccessibleFemale: String?
    let AccessibleUnisex: String?
    let Unisex:           String?
    let PayToilet:        String?
    let BabyChange:       String?
    let Notes:            String?
    let OpeningHours:     String?
}

// MARK: - Refuge Restrooms JSON model

private struct RefugeEntry: Decodable {
    let id:         Int
    let name:       String
    let street:     String?
    let city:       String?
    let state:      String?
    let country:    String?
    let latitude:   Double
    let longitude:  Double
    let accessible: Bool
    let directions: String?
}
